org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java

```diff
@@ -116,22 +116,30 @@ public void generateCode(BlockScope currentScope, CodeStream codeStream, boolean
 		codeStream.addVariable(this.secretExpressionValue);
 	}
 	codeStream.instance_of(this.type, this.type.resolvedType);
-	if (this.elementVariable != null) {
-		BranchLabel actionLabel = new BranchLabel(codeStream);
-		codeStream.dup();
-		codeStream.ifeq(actionLabel);
+	if (this.pattern != null) {
+		BranchLabel falseLabel = new BranchLabel(codeStream);
+		BranchLabel trueLabel = new BranchLabel(codeStream);
+		BranchLabel continueLabel = new BranchLabel(codeStream);
+		codeStream.ifeq(falseLabel);
+
 		if (this.secretExpressionValue != null) {
 			codeStream.load(this.secretExpressionValue);
 			codeStream.removeVariable(this.secretExpressionValue);
 		} else {
 			this.expression.generateCode(currentScope, codeStream, true);
 		}
-		codeStream.checkcast(this.type, this.type.resolvedType, codeStream.position);
-		this.elementVariable.binding.recordInitializationStartPC(codeStream.position);
-		codeStream.store(this.elementVariable.binding, false);
-		codeStream.removeVariable(this.elementVariable.binding);
-		codeStream.recordPositionsFrom(codeStream.position, this.sourceEnd);
-		actionLabel.place();
+		this.pattern.generateOptimizedBoolean(currentScope, codeStream, trueLabel, falseLabel);
+
+		for (LocalVariableBinding patternVariableBinding : this.pattern.bindingsWhenTrue()) {
+			codeStream.removeVariable(patternVariableBinding);
+		}
```

Comment:
This is wrong code. You are perhaps following the problematic coding pattern prevalent in the code base that I am trying to weed out. In pattern matching implementation, variables are added and removed without much rhyme or reason. In this place, instanceof is true and the pattern binding is live - it should not be removed. The flow analysis should determine when it becomes not definitely assigned and close the ranges with a call to record to `org.eclipse.jdt.internal.compiler.lookup.LocalVariableBinding.recordInitializationEndPC(int)`

Reply:
Note however the call to `codeStream.removeVariable(this.secretExpressionValue);` just above in line 127 is correct. We are done with the use of secret variable and it can be done away with. It being a compiler injected variable, its scope and live range is entirely upto us. We want to keep it alive for the minimum range required.

But pattern bindings are programmer defined variables and their scope and live ranges (pc intervals where are they are definitely assigned per flow scope) are spelled out by JLS. analyzeCode() phase computes the definite assignment ranges for locals.